### DISCUSSION TITLE
convert multiple images to one pdf every image one page and centered

### DISCUSSION THREAD

**Discussion by SabineMa** [ImageMagick/ImageMagick#2784] [Help] [answered]
Hi,

I have a problem that seems to be easy but I don't get it.
I have a list of N pngs, each one with different size (that I do not know before). I create the result pdf with:

convert {sourcefileslist} -page a4 resultfile.pdf

Until here everything is fine. Only one thing is not nice: small images are placed in the lower left edge of the page.  This is the case in this example on page 4 and on the last page. Please ignore that the small image is cropped. 

[step1.pdf](https://github.com/ImageMagick/ImageMagick/files/5470934/step1.pdf)

What I want is that the files are placed in the center of each page, also the small ones. For this, I add -gravity like this:

convert {sourcefileslist}   -page a4   -gravity center   resultfile.pdf

Now the small file is ok - centered but the other files are not shown completely.

Note that the input file sizes are not known by me before. 
This is just an example. 

[step2.pdf](https://github.com/ImageMagick/ImageMagick/files/5470945/step2.pdf)

I do not get it how to do this and ask for a hint to the community.

Sabine

**Comment by fmw42** [comment]
What version of Imagemagick and on what platform? Please note that -page does not resize. If you need it to resize, then you will have to use -resize WxH where WxH are pixel dimensions corresponding to A4 (except on IM 7.0.10.35 with the new papersize string format). If you just want to place in the center of an A4 background, then

`convert <image list> -gravity center -background white -extent 595x842 x.pdf`

On IM 7.0.10.35, one can now do:

`magick <image list> -gravity center -background white -extent "%[papersize:A4]" x.pdf`

**Reply by SabineMa** [reply]
Hi, 

thanks a lot for your answer. I use ImageMagick 7.0.10-35 Q16 x86_64 2020-11-01  iOS.

When I try with:
`magick <image list> -gravity center -background white -extent "%[papersize:A4]" x.pdf`
the result is better because it is centered but the larger files are not resized: 
[202010311045-337841000.pdf](https://github.com/ImageMagick/ImageMagick/files/5471567/202010311045-337841000.pdf)

Then I tried with:

`magick <image list> -gravity center -background white -resize 595x842 -extent "%[papersize:A4]" x.pdf`
which leads to loosing the original size of the smaller image (last page).
[202010311045-337841000.pdf](https://github.com/ImageMagick/ImageMagick/files/5471577/202010311045-337841000.pdf)
I don't want the smaller images to be extended, only centered.

Do you have an idea what's wrong here?

Regards 
Sabine

**Comment by fmw42** [comment, marked as answer]
try

`magick <image list> -resize "%[papersize:A4]>" -gravity center -background white -extent "%[papersize:A4]" x.pdf`

the > tell the resize to only resize if the image is larger than A4

see  https://imagemagick.org/script/command-line-processing.php#geometry

**Reply by SabineMa** [reply]
I am very happy now, thanks a lot for your great help! 

The result is exactly what I need:
[202010311045-337841000.pdf](https://github.com/ImageMagick/ImageMagick/files/5471654/202010311045-337841000.pdf)

made my day 👍 
Sabine

**Reply by tobx** [reply]
> `magick <image list> -resize "%[papersize:A4]>" -gravity center -background white -extent "%[papersize:A4]" x.pdf`

Technically that does not center the image in the PDF, but it extends the image to A4 with a white border and then places it on a PDF page. The resulting image is larger in size especially for small images with high resolutions. I have yet to find a way to really center an image on a PDF page.

**Reply by fmw42** [reply]
@tobx. It seems to work for me in IM 7.1.1.47

magick logo: lena.png barn.jpg monet2.jpg -resize "%[papersize:A4]>" -gravity center -background white -extent "%[papersize:A4]" x.pdf

[x.pdf](https://github.com/user-attachments/files/19722268/x.pdf)

**Reply by tobx** [reply]
> @tobx. It seems to work for me in IM 7.1.1.47
> 
> magick logo: lena.png barn.jpg monet2.jpg -resize "%[papersize:A4]>" -gravity center -background white -extent "%[papersize:A4]" x.pdf
> 
> [x.pdf](https://github.com/user-attachments/files/19722268/x.pdf)

I think I was not clear enough. With "The resulting image is larger in size" I meant file size not PDF dimensions. You can simply check this by converting a single image with your command like this (converting it to `png` instead of `pdf`:

`magick lena.png -resize "%[papersize:A4]>" -gravity center -background white -extent "%[papersize:A4]" out.png`

The resulting image has white borders and therefore is larger in file size.

So your initial command places an image resized to A4 (by adding white borders) in the corner of a PDF page. The correct or optimal solution would be to not resize the image itself, but just place it in the center of a A4 PDF page. It is possible with the `-page +x+y` offset, but then you need to know the image dimensions before.

**Comment by snibgo** [comment]
When IM writes raster images to PDF, each page will be the same size as the raster image. There is no way to create a blank PDF margin around the raster images. Instead, as fmw42 shows, we can extend the raster image with a blank margin, and then write that larger raster image to PDF.

**Reply by tobx** [reply]
Actually you can make the PDF page larger than the image with the `-page` option.

**Comment by snibgo** [comment]
Thanks for the correction, @tobx. Yes, `-page` does the job. For example, we make a blue raster image 267x233 and place it in the centre of a larger PDF page:
```
magick -page 1200x600+%[fx:(1200-267)/2]+%[fx:(600-233)/2] -size 267x233 xc:Blue x.pdf
```
`pdfimages -list x.pdf` demonstrates that the raster image in x.pdf is only 267x233 pixels.

**Reply by tobx** [reply]
Nice, I did not know that you can do math within the command. Still the issue with this is, that you cannot do batch processing for multiple images with different sizes or create multi page PDFs.

**Comment by snibgo** [comment]
>  ... batch processing for multiple images with different sizes or create multi page PDFs.

Suppose we have two input raster images: toes.png and a small blue image. We want a PDF with two pages, where each input image is centred on its own page.
```
magick toes.png -size 20x30 xc:Blue -set page 1200x600+%[fx:(1200-w)/2]+%[fx:(600-h)/2] -density 72 -units PixelsPerInch x.pdf
```

**Reply by tobx** [reply]
Oh wow, you can use `w` and `h` as variables, awesome, thank you!

**Comment by snibgo** [comment]
Yes. My command above sets page offsets for each image, using appropriate values for `w` and `h`, the image width and height. For other variables that can be used, see https://imagemagick.org/script/escape.php . For fx, see the official page https://imagemagick.org/script/fx.php and my page [New FX](http://im.snibgo.com/fxnew.htm).

My command also sets all the images to 72 dpi, because that is the usual resolution of PDF files. If that isn't done, then any images that are not 72 dpi will be placed at weird offsets. An alternative might be to modify the offsets, perhaps multiply by 72 and divide by the image resolution, or something.

**Reply by tobx** [reply]
`multiply by 72 and divide by the image resolution` works, I just tried and finally was able to create my PDF even with the correct dimensions in centimeters.
